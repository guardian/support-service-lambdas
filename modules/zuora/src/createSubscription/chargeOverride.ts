import type { ProductCatalog } from '@modules/product-catalog/productCatalog';
import { ProductPurchase } from '@modules/product-catalog/productPurchaseSchema';

export const getChargeOverride = (
	productCatalog: ProductCatalog,
	productPurchase: ProductPurchase,
): { productRatePlanChargeId: string; overrideAmount: number } | undefined => {
	if (productPurchase.product === 'Contribution') {
		const chargeId =
			productCatalog.Contribution.ratePlans[productPurchase.ratePlan].charges
				.Contribution.id;
		return {
			productRatePlanChargeId: chargeId,
			overrideAmount: productPurchase.amount,
		};
	} else if (productPurchase.product === 'SupporterPlus') {
		if (
			// TODO: Is this right?
			//These are the only rate plans that have a contribution charge
			productPurchase.ratePlan === 'Annual' ||
			productPurchase.ratePlan === 'Monthly'
		) {
			const chargeId =
				productCatalog.SupporterPlus.ratePlans[productPurchase.ratePlan].charges
					.Contribution.id;
			return {
				productRatePlanChargeId: chargeId,
				overrideAmount: productPurchase.amount,
			};
		}
	}
	return;
};
