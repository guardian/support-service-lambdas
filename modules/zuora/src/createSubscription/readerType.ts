export enum ReaderType {
	Direct = 'Direct',
	Gift = 'Gift',
	Patron = 'Patron',
}
