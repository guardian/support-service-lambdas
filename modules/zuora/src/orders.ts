export type ProcessingOptions = {
	runBilling: boolean;
	collectPayment: boolean;
};
export type PreviewOptions = {
	previewThruType: 'SpecificDate';
	previewTypes: ['BillingDocs'];
	specificPreviewThruDate: string;
};
export type OrderActionType =
	| 'ChangePlan'
	| 'TermsAndConditions'
	| 'RenewSubscription'
	| 'UpdateProduct'
	| 'AddProduct';

type BaseOrderAction = {
	type: OrderActionType;
	triggerDates: [
		{
			name: 'ContractEffective';
			triggerDate: string;
		},
		{
			name: 'ServiceActivation';
			triggerDate: string;
		},
		{
			name: 'CustomerAcceptance';
			triggerDate: string;
		},
	];
};
export type ChangePlanOrderAction = BaseOrderAction & {
	type: 'ChangePlan';
	changePlan: {
		productRatePlanId: string;
		subType: 'Upgrade';
		newProductRatePlan: {
			productRatePlanId: string;
			chargeOverrides: [
				{
					productRatePlanChargeId: string;
					pricing: {
						recurringFlatFee: {
							listPrice: number;
						};
					};
				},
			];
		};
	};
};
export type DiscountOrderAction = BaseOrderAction & {
	type: 'AddProduct';
	addProduct: {
		productRatePlanId: string;
	};
};

export type UpdateProductOrderAction = BaseOrderAction & {
	type: 'UpdateProduct';
	updateProduct: {
		ratePlanId: string;
		chargeUpdates: [
			{
				chargeNumber: string;
				pricing: {
					recurringFlatFee: {
						listPrice: number;
					};
				};
			},
		];
	};
};
export type RenewSubscriptionOrderAction = BaseOrderAction & {
	type: 'RenewSubscription';
	renewSubscription: object;
};
export type TermsAndConditionsOrderAction = BaseOrderAction & {
	type: 'TermsAndConditions';
	termsAndConditions: {
		lastTerm: {
			termType: 'TERMED';
			endDate: string;
		};
	};
};
export type OrderAction =
	| ChangePlanOrderAction
	| RenewSubscriptionOrderAction
	| TermsAndConditionsOrderAction
	| UpdateProductOrderAction
	| DiscountOrderAction;

export type OrderRequest = {
	orderDate: string;
	existingAccountNumber: string;
	description?: string;
	subscriptions: Array<{
		subscriptionNumber: string;
		customFields?: { LastPlanAddedDate__c?: string };
		orderActions: OrderAction[];
	}>;
};
export type PreviewOrderRequest = OrderRequest & {
	previewOptions: PreviewOptions;
};
export type CreateOrderRequest = OrderRequest & {
	processingOptions: ProcessingOptions;
};
