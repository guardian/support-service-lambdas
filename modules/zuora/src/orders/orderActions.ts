import type { BillingPeriod } from '@modules/billingPeriod';
import type { TermType } from '@modules/product-catalog/productCatalog';
import type { Dayjs } from 'dayjs';
import { zuoraDateFormat } from '../utils/common';

export type TriggerDates = [
	{
		name: 'ContractEffective';
		triggerDate: string;
	},
	{
		name: 'CustomerAcceptance';
		triggerDate: string;
	},
];

export type OrderActionType =
	| 'ChangePlan'
	| 'TermsAndConditions'
	| 'RenewSubscription'
	| 'UpdateProduct'
	| 'CreateSubscription'
	| 'AddProduct';

type BaseOrderAction = {
	type: OrderActionType;
	triggerDates: TriggerDates;
};

export type ChangePlanOrderAction = BaseOrderAction & {
	type: 'ChangePlan';
	changePlan: {
		productRatePlanId: string;
		subType: 'Upgrade';
		newProductRatePlan: {
			productRatePlanId: string;
			chargeOverrides: [
				{
					productRatePlanChargeId: string;
					pricing: {
						recurringFlatFee: {
							listPrice: number;
						};
					};
				},
			];
		};
	};
};
export type DiscountOrderAction = BaseOrderAction & {
	type: 'AddProduct';
	addProduct: {
		productRatePlanId: string;
	};
};
export type UpdateProductOrderAction = BaseOrderAction & {
	type: 'UpdateProduct';
	updateProduct: {
		ratePlanId: string;
		chargeUpdates: [
			{
				chargeNumber: string;
				pricing: {
					recurringFlatFee: {
						listPrice: number;
					};
				};
			},
		];
	};
};
export type RenewSubscriptionOrderAction = BaseOrderAction & {
	type: 'RenewSubscription';
};
export type TermsAndConditionsOrderAction = BaseOrderAction & {
	type: 'TermsAndConditions';
	termsAndConditions: {
		lastTerm: {
			termType: 'TERMED';
			endDate: string;
		};
	};
};
export type CreateSubscriptionOrderAction = BaseOrderAction & {
	type: 'CreateSubscription';
	createSubscription: {
		terms: {
			autoRenew: boolean;
			initialTerm: {
				period: number;
				periodType: BillingPeriod;
				termType: 'TERMED';
			};
			renewalSetting: 'RENEW_WITH_SPECIFIC_TERM';
			renewalTerms: Array<{
				period: number;
				periodType: BillingPeriod;
			}>;
		};
		subscribeToRatePlans: [
			{
				productRatePlanId: string;
				chargeOverrides?: Array<{
					productRatePlanChargeId: string;
					pricing: {
						recurringFlatFee: {
							listPrice: number;
						};
					};
				}>;
			},
		];
	};
	customFields?: {
		DeliveryAgent__c: string | undefined;
		ReaderType__c: string;
	};
};
export type OrderAction =
	| ChangePlanOrderAction
	| RenewSubscriptionOrderAction
	| TermsAndConditionsOrderAction
	| UpdateProductOrderAction
	| DiscountOrderAction
	| CreateSubscriptionOrderAction;

export function singleTriggerDate(applyFromDate: Dayjs): TriggerDates {
	return [
		{
			name: 'ContractEffective',
			triggerDate: zuoraDateFormat(applyFromDate),
		},
		{
			name: 'CustomerAcceptance',
			triggerDate: zuoraDateFormat(applyFromDate),
		},
	];
}

export function initialTermInDays(
	contractEffectiveDate: Dayjs,
	customerAcceptanceDate: Dayjs,
	termLength: number,
) {
	// This functionality is ported over from the previous version of this code.
	// I think the reason we need the term in days is because for delivery products
	// the term starts on the contract effective date (when the order is placed)
	// but the customer acceptance date (when the first delivery is made) can be
	// some days later. We need to ensure that the term covers the full period
	// from contract effective date to [termLength] after customer acceptance date.
	const termEnd = customerAcceptanceDate.add(termLength, 'month');
	return termEnd.diff(contractEffectiveDate, 'day');
}

// Builder function to simplify the creation of a CreateSubscriptionOrderAction
// object as a lot of it is boilerplate.
export function buildCreateSubscriptionOrderAction({
	productRatePlanId,
	contractEffectiveDate,
	customerAcceptanceDate,
	chargeOverride,
	termType,
	termLengthInMonths,
}: {
	productRatePlanId: string;
	contractEffectiveDate: Dayjs;
	customerAcceptanceDate: Dayjs;
	chargeOverride?: { productRatePlanChargeId: string; overrideAmount: number };
	termType: TermType;
	termLengthInMonths: number;
}): CreateSubscriptionOrderAction {
	const chargeOverrides = chargeOverride
		? [
				{
					productRatePlanChargeId: chargeOverride.productRatePlanChargeId,
					pricing: {
						recurringFlatFee: {
							listPrice: chargeOverride.overrideAmount,
						},
					},
				},
			]
		: [];

	const [initialPeriodLength, initialPeriodType, autoRenew] =
		termType === 'Recurring'
			? [12, 'Month', true]
			: [
					initialTermInDays(
						contractEffectiveDate,
						customerAcceptanceDate,
						termLengthInMonths,
					),
					'Days',
					false,
				];

	const terms = {
		autoRenew: autoRenew,
		initialTerm: {
			period: initialPeriodLength,
			periodType: initialPeriodType as BillingPeriod,
			termType: 'TERMED' as const,
		},
		renewalSetting: 'RENEW_WITH_SPECIFIC_TERM' as const,
		renewalTerms: [
			{
				period: termLengthInMonths,
				periodType: 'Month' as BillingPeriod,
			},
		],
	};

	return {
		type: 'CreateSubscription',
		triggerDates: [
			{
				name: 'ContractEffective',
				triggerDate: zuoraDateFormat(contractEffectiveDate),
			},
			{
				name: 'CustomerAcceptance',
				triggerDate: zuoraDateFormat(customerAcceptanceDate),
			},
		],
		createSubscription: {
			terms: terms,
			subscribeToRatePlans: [
				{
					productRatePlanId,
					chargeOverrides,
				},
			],
		},
	};
}
