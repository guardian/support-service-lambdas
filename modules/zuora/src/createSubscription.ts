import type { ZuoraClient } from '@modules/zuora/zuoraClient';
import { OrderRequest } from '@modules/zuora/orders/orders';
import { z } from 'zod';

const createSubscriptionResponseSchema = z.object({
	orderNumber: z.string(),
	accountNumber: z.string(),
	subscriptionNumbers: z.array(z.string()),
});

type CreateSubscriptionResponse = z.infer<
	typeof createSubscriptionResponseSchema
>;

export const createSubscription = async (
	zuoraClient: ZuoraClient,
	createSubscriptionRequest: OrderRequest,
): Promise<CreateSubscriptionResponse> => {
	const path = `/v1/orders`;
	const body = JSON.stringify(createSubscriptionRequest);
	return zuoraClient.post(path, body, createSubscriptionResponseSchema, {
		'zuora-version': '211.0',
	});
};
