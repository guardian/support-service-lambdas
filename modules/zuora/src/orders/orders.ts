import { OrderAction } from '@modules/zuora/orders/orderActions';
import { NewAccount } from '@modules/zuora/orders/newAccount';

export type ProcessingOptions = {
	runBilling: boolean;
	collectPayment: boolean;
};
export type PreviewOptions = {
	previewThruType: 'SpecificDate';
	previewTypes: ['BillingDocs'];
	specificPreviewThruDate: string;
};

type NewAccountOrderRequest = {
	newAccount: NewAccount;
};
export type ExistingAccountOrderRequest = {
	existingAccountNumber: string;
};
type AccountOrderRequest = NewAccountOrderRequest | ExistingAccountOrderRequest;
export type OrderRequest = AccountOrderRequest & {
	orderDate: string;
	description?: string;
	subscriptions: Array<{
		subscriptionNumber?: string;
		orderActions: OrderAction[];
	}>;
};
export type PreviewOrderRequest = OrderRequest & {
	previewOptions: PreviewOptions;
};
export type CreateOrderRequest = OrderRequest & {
	processingOptions: ProcessingOptions;
};
