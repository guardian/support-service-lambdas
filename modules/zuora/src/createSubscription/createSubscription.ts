import type { IsoCurrency } from '@modules/internationalisation/currency';
import { ProductCatalog } from '@modules/product-catalog/productCatalog';
import type { ProductPurchase } from '@modules/product-catalog/productPurchaseSchema';
import dayjs from 'dayjs';
import { z } from 'zod';
import { getChargeOverride } from '@modules/zuora/createSubscription/chargeOverride';
import { getSubscriptionDates } from '@modules/zuora/createSubscription/subscriptionDates';
import type { Contact } from '@modules/zuora/orders/newAccount';
import { buildNewAccountObject } from '@modules/zuora/orders/newAccount';
import { buildCreateSubscriptionOrderAction } from '@modules/zuora/orders/orderActions';
import type { AnyOrderRequest } from '@modules/zuora/orders/orderRequests';
import { executeOrderRequest } from '@modules/zuora/orders/orderRequests';
import type {
	PaymentGateway,
	PaymentMethod,
} from '@modules/zuora/orders/paymentMethods';
import { zuoraDateFormat } from '@modules/zuora/utils';
import type { ZuoraClient } from '@modules/zuora/zuoraClient';
import { getProductRatePlanId } from '@modules/zuora/createSubscription/getProductRatePlanId';

const createSubscriptionResponseSchema = z.object({
	orderNumber: z.string(),
	accountNumber: z.string(),
	subscriptionNumbers: z.array(z.string()),
	invoiceNumbers: z.array(z.string()).optional(),
	paymentNumber: z.string().optional(),
	paidAmount: z.number().optional(),
});

export type CreateSubscriptionResponse = z.infer<
	typeof createSubscriptionResponseSchema
>;

type CreateSubscriptionInputFields<T extends PaymentMethod> = {
	accountName: string;
	createdRequestId: string;
	salesforceAccountId: string;
	salesforceContactId: string;
	identityId: string;
	currency: IsoCurrency;
	paymentGateway: PaymentGateway<T>;
	paymentMethod: T;
	billToContact: Contact;
	productPurchase: ProductPurchase;
	runBilling?: boolean;
	collectPayment?: boolean;
};

function buildCreateSubscriptionRequest<T extends PaymentMethod>(
	productCatalog: ProductCatalog,
	{
		accountName,
		createdRequestId,
		salesforceAccountId,
		salesforceContactId,
		identityId,
		currency,
		paymentGateway,
		paymentMethod,
		billToContact,
		productPurchase,
		runBilling,
		collectPayment,
	}: CreateSubscriptionInputFields<T>,
): AnyOrderRequest {
	// TODO:
	//  Apply promotion if present
	//  ReaderType - GIFT, PATRON
	//  DeliveryAgent âœ…
	//  Validate paper payment gateway
	//  Set term & autoRenew correctly for GW gifts (and S+ students?)
	//  Set contribution amount correctly for S+ (amount - cost)
	//  Output state
	//  CSR mode is NOT needed
	const { soldToContact, deliveryAgent, deliveryInstructions } = {
		soldToContact: undefined,
		deliveryAgent: '',
		deliveryInstructions: undefined,
		...productPurchase,
	};

	const newAccount = buildNewAccountObject({
		accountName: accountName,
		createdRequestId: createdRequestId,
		salesforceAccountId: salesforceAccountId,
		salesforceContactId: salesforceContactId,
		identityId: identityId,
		currency: currency,
		paymentGateway: paymentGateway,
		paymentMethod: paymentMethod,
		billToContact: billToContact,
		soldToContact: soldToContact,
		deliveryInstructions: deliveryInstructions,
	});
	const { contractEffectiveDate, customerAcceptanceDate } =
		getSubscriptionDates(dayjs(), productPurchase);

	const chargeOverride = getChargeOverride(productCatalog, productPurchase);

	const createSubscriptionOrderAction = buildCreateSubscriptionOrderAction({
		productRatePlanId: getProductRatePlanId(productCatalog, productPurchase),
		contractEffectiveDate: contractEffectiveDate,
		customerAcceptanceDate: customerAcceptanceDate,
		chargeOverride: chargeOverride,
		deliveryAgent: deliveryAgent.toString(),
	});
	return {
		newAccount: newAccount,
		orderDate: zuoraDateFormat(dayjs()),
		description: `Created by createSubscription.ts in support-service-lambdas`,
		subscriptions: [
			{
				orderActions: [createSubscriptionOrderAction],
			},
		],
		processingOptions: {
			runBilling: runBilling ?? true,
			collectPayment: collectPayment ?? true,
		},
	};
}
export const createSubscription = async <T extends PaymentMethod>(
	zuoraClient: ZuoraClient,
	productCatalog: ProductCatalog,
	inputFields: CreateSubscriptionInputFields<T>,
): Promise<CreateSubscriptionResponse> => {
	return executeOrderRequest(
		zuoraClient,
		buildCreateSubscriptionRequest(productCatalog, inputFields),
		createSubscriptionResponseSchema,
	);
};
