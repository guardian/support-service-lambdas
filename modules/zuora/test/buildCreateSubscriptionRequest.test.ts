import {
	buildCreateSubscriptionRequest,
	CreateSubscriptionInputFields,
} from '../src/createSubscription/createSubscription';
import { ReaderType } from '../src/createSubscription/readerType';
import { AppliedPromotion, Promotion } from '@modules/promotions/schema';
import dayjs from 'dayjs';
import { IsoCurrency } from '@modules/internationalisation/currency';
import { Stage } from '@modules/stage';
import {
	CreditCardReferenceTransaction,
	PaymentGateway,
} from '@modules/zuora/orders/paymentMethods';
import { CreateSubscriptionOrderAction } from '@modules/zuora/orders/orderActions';
import code from '../../zuora-catalog/test/fixtures/catalog-code.json';
import { generateProductCatalog } from '@modules/product-catalog/generateProductCatalog';
import { SupportRegionId } from '@modules/internationalisation/countryGroup';

const contractEffectiveDate = dayjs('2025-09-01');
const customerAcceptanceDate = dayjs('2025-09-05');

jest.mock('../src/createSubscription/subscriptionDates', () => ({
	getSubscriptionDates: jest.fn(() => ({
		contractEffectiveDate: dayjs(contractEffectiveDate),
		customerAcceptanceDate: dayjs(customerAcceptanceDate),
	})),
}));

const productCatalog = generateProductCatalog(code);
const promotions: Promotion[] = [
	{
		name: 'Test Promo',
		promotionType: { name: 'percent_discount', amount: 10, durationMonths: 3 },
		appliesTo: {
			countries: new Set(['GB']),
			productRatePlanIds: new Set(['8ad08cbd8586721c01858804e3275376']),
		},
		codes: { 'Channel 0': ['PROMO10'] },
		starts: new Date('2024-01-01'),
		expires: new Date('2099-01-01'),
	},
	{
		name: 'Patron Promo',
		promotionType: { name: 'percent_discount', amount: 10, durationMonths: 3 },
		appliesTo: {
			countries: new Set(['GB']),
			productRatePlanIds: new Set(['8ad08cbd8586721c01858804e3275376']),
		},
		codes: { 'Channel 0': ['TEST_PATRON'] },
		starts: new Date('2024-01-01'),
		expires: new Date('2099-01-01'),
	},
];

const paymentMethod = {
	type: 'CreditCardReferenceTransaction' as const,
	tokenId: 'test_token_id',
	secondTokenId: 'test_second_token_id',
	cardNumber: '42424242424242',
	cardType: 'Visa',
	expirationMonth: 12,
	expirationYear: 2099,
};

const baseStripeInput = {
	stage: 'CODE' as Stage,
	accountName: 'Test Account',
	createdRequestId: 'request-123',
	salesforceAccountId: 'sf-acc-1',
	salesforceContactId: 'sf-con-1',
	identityId: 'id-1',
	currency: 'GBP' as IsoCurrency,
	paymentGateway:
		'Stripe PaymentIntents GNM Membership' as PaymentGateway<CreditCardReferenceTransaction>,
	paymentMethod: paymentMethod,
	billToContact: {
		firstName: 'A',
		lastName: 'B',
		workEmail: 'test@test.com',
		country: 'GB',
	},
};

const supporterPlusInput: CreateSubscriptionInputFields<CreditCardReferenceTransaction> =
	{
		...baseStripeInput,
		productPurchase: {
			product: 'SupporterPlus',
			ratePlan: 'Monthly',
			amount: 12,
		},
	};
const guardianWeeklyGiftInput: CreateSubscriptionInputFields<CreditCardReferenceTransaction> =
	{
		...baseStripeInput,
		productPurchase: {
			product: 'GuardianWeeklyDomestic',
			ratePlan: 'OneYearGift',
			firstDeliveryDate: dayjs().add(1, 'month').toDate(),
			deliveryContact: {
				firstName: 'G',
				lastName: 'W',
				workEmail: 'test@test.com',
				country: 'GB',
				address1: '90 York Way',
				city: 'London',
				postalCode: 'N19GU',
			},
		},
	};

const isCreateSubscriptionOrderAction = (
	obj: any,
): obj is CreateSubscriptionOrderAction => {
	return (
		obj &&
		typeof obj === 'object' &&
		'type' in obj &&
		obj.type === 'CreateSubscription'
	);
};

describe('the buildCreateSubscriptionRequest function', () => {
	it('builds request without promotion or gift', () => {
		const request = buildCreateSubscriptionRequest(
			productCatalog,
			promotions,
			supporterPlusInput,
		);
		if (!('newAccount' in request)) {
			throw new Error('Expected newAccount in request');
		}
		expect(request.newAccount.name).toEqual('Test Account');
		expect(request.subscriptions[0]?.orderActions[0]?.type).toBe(
			'CreateSubscription',
		);
		expect(request.subscriptions[0]?.customFields?.ReaderType__c).toBe(
			ReaderType.Direct,
		);
		expect(request.processingOptions.runBilling).toBe(true);
		expect(request.processingOptions.collectPayment).toBe(true);
		expect(request).toMatchSnapshot();
	});

	it('builds request with applied promotion', () => {
		const input = {
			...supporterPlusInput,
			appliedPromotion: {
				promoCode: 'PROMO10',
				supportRegionId: SupportRegionId.UK,
			} as AppliedPromotion,
		};
		const request = buildCreateSubscriptionRequest(
			productCatalog,
			promotions,
			input,
		);
		const orderAction = request.subscriptions[0]?.orderActions[0];
		if (!isCreateSubscriptionOrderAction(orderAction)) {
			throw new Error('Expected CreateOrderRequest');
		}

		expect(orderAction.createSubscription.subscribeToRatePlans.length).toBe(2);
		expect(request.subscriptions[0]?.customFields?.ReaderType__c).toBe(
			ReaderType.Direct,
		);
		expect(request).toMatchSnapshot();
	});

	it('builds request with gift recipient', () => {
		const input = {
			...guardianWeeklyGiftInput,
			giftRecipient: {
				email: 'gift@domain.com',
				firstName: 'Gift',
				lastName: 'Recipient',
			},
		};
		const request = buildCreateSubscriptionRequest(
			productCatalog,
			promotions,
			input,
		);
		expect(request.subscriptions[0]?.customFields?.ReaderType__c).toBe(
			ReaderType.Gift,
		);
		expect(request).toMatchSnapshot();
	});

	it('sets the correct ReaderType for a patron promotion', () => {
		const input = {
			...supporterPlusInput,
			appliedPromotion: {
				promoCode: 'TEST_PATRON',
				supportRegionId: SupportRegionId.UK,
			} as AppliedPromotion,
		};
		const request = buildCreateSubscriptionRequest(
			productCatalog,
			promotions,
			input,
		);
		expect(request.subscriptions[0]?.customFields?.ReaderType__c).toBe(
			ReaderType.Patron,
		);
	});

	it('sets runBilling and collectPayment to false if specified', () => {
		const input = {
			...supporterPlusInput,
			runBilling: false,
			collectPayment: false,
		};
		const request = buildCreateSubscriptionRequest(
			productCatalog,
			promotions,
			input,
		);
		expect(request.processingOptions.runBilling).toBe(false);
		expect(request.processingOptions.collectPayment).toBe(false);
	});
});
