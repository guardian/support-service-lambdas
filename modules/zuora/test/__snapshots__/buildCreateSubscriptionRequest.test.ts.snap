// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCreateSubscriptionRequest builds request with applied promotion 1`] = `
{
  "description": "Created by createSubscription.ts in support-service-lambdas",
  "newAccount": {
    "autoPay": true,
    "billCycleDay": 0,
    "billToContact": {
      "country": "GB",
      "firstName": "A",
      "lastName": "B",
      "workEmail": "test@test.com",
    },
    "crmId": "sf-acc-1",
    "currency": "GBP",
    "customFields": {
      "CreatedRequestId__c": "request-123",
      "IdentityId__c": "id-1",
      "sfContactId__c": "sf-con-1",
    },
    "name": "Test Account",
    "paymentGateway": "Stripe PaymentIntents GNM Membership",
    "paymentMethod": {
      "cardNumber": "42424242424242",
      "cardType": "Visa",
      "expirationMonth": 12,
      "expirationYear": 2099,
      "secondTokenId": "test_second_token_id",
      "tokenId": "test_token_id",
      "type": "CreditCardReferenceTransaction",
    },
    "soldToContact": undefined,
  },
  "orderDate": "2025-09-17",
  "processingOptions": {
    "collectPayment": true,
    "runBilling": true,
  },
  "subscriptions": [
    {
      "customFields": {
        "DeliveryAgent__c": "",
        "ReaderType__c": "Direct",
      },
      "orderActions": [
        {
          "createSubscription": {
            "subscribeToRatePlans": [
              {
                "chargeOverrides": [
                  {
                    "pricing": {
                      "recurringFlatFee": {
                        "listPrice": 0,
                      },
                    },
                    "productRatePlanChargeId": "8ad09ea0858682bb0185880ac57f4c4c",
                  },
                ],
                "productRatePlanId": "8ad08cbd8586721c01858804e3275376",
              },
              {
                "chargeOverrides": [
                  {
                    "endDate": {
                      "endDateCondition": "Fixed_Period",
                      "upToPeriods": 3,
                      "upToPeriodsType": "Months",
                    },
                    "pricing": {
                      "discount": {
                        "discountPercentage": 10,
                      },
                    },
                    "productRatePlanChargeId": "2c92c0f957220b5d0157299c97a60bbd",
                  },
                ],
                "productRatePlanId": "2c92c0f85721ff7c01572942235b6d7a",
              },
            ],
            "terms": {
              "autoRenew": true,
              "initialTerm": {
                "period": 12,
                "periodType": "Month",
                "termType": "TERMED",
              },
              "renewalSetting": "RENEW_WITH_SPECIFIC_TERM",
              "renewalTerms": [
                {
                  "period": 12,
                  "periodType": "Month",
                },
              ],
            },
          },
          "triggerDates": [
            {
              "name": "ContractEffective",
              "triggerDate": "2025-09-01",
            },
            {
              "name": "CustomerAcceptance",
              "triggerDate": "2025-09-05",
            },
          ],
          "type": "CreateSubscription",
        },
      ],
    },
  ],
}
`;

exports[`buildCreateSubscriptionRequest builds request with gift recipient 1`] = `
{
  "description": "Created by createSubscription.ts in support-service-lambdas",
  "newAccount": {
    "autoPay": true,
    "billCycleDay": 0,
    "billToContact": {
      "country": "GB",
      "firstName": "A",
      "lastName": "B",
      "workEmail": "test@test.com",
    },
    "crmId": "sf-acc-1",
    "currency": "GBP",
    "customFields": {
      "CreatedRequestId__c": "request-123",
      "IdentityId__c": "id-1",
      "sfContactId__c": "sf-con-1",
    },
    "name": "Test Account",
    "paymentGateway": "Stripe PaymentIntents GNM Membership",
    "paymentMethod": {
      "cardNumber": "42424242424242",
      "cardType": "Visa",
      "expirationMonth": 12,
      "expirationYear": 2099,
      "secondTokenId": "test_second_token_id",
      "tokenId": "test_token_id",
      "type": "CreditCardReferenceTransaction",
    },
    "soldToContact": {
      "SpecialDeliveryInstructions__c": undefined,
      "address1": "90 York Way",
      "city": "London",
      "country": "GB",
      "firstName": "G",
      "lastName": "W",
      "postalCode": "N19GU",
      "workEmail": "test@test.com",
    },
  },
  "orderDate": "2025-09-17",
  "processingOptions": {
    "collectPayment": true,
    "runBilling": true,
  },
  "subscriptions": [
    {
      "customFields": {
        "DeliveryAgent__c": "",
        "ReaderType__c": "Gift",
      },
      "orderActions": [
        {
          "createSubscription": {
            "subscribeToRatePlans": [
              {
                "chargeOverrides": [],
                "productRatePlanId": "2c92c0f867cae0700167eff921734f7b",
              },
            ],
            "terms": {
              "autoRenew": false,
              "initialTerm": {
                "period": 369,
                "periodType": "Day",
                "termType": "TERMED",
              },
              "renewalSetting": "RENEW_WITH_SPECIFIC_TERM",
              "renewalTerms": [
                {
                  "period": 12,
                  "periodType": "Month",
                },
              ],
            },
          },
          "triggerDates": [
            {
              "name": "ContractEffective",
              "triggerDate": "2025-09-01",
            },
            {
              "name": "CustomerAcceptance",
              "triggerDate": "2025-09-05",
            },
          ],
          "type": "CreateSubscription",
        },
      ],
    },
  ],
}
`;

exports[`buildCreateSubscriptionRequest builds request without promotion or gift 1`] = `
{
  "description": "Created by createSubscription.ts in support-service-lambdas",
  "newAccount": {
    "autoPay": true,
    "billCycleDay": 0,
    "billToContact": {
      "country": "GB",
      "firstName": "A",
      "lastName": "B",
      "workEmail": "test@test.com",
    },
    "crmId": "sf-acc-1",
    "currency": "GBP",
    "customFields": {
      "CreatedRequestId__c": "request-123",
      "IdentityId__c": "id-1",
      "sfContactId__c": "sf-con-1",
    },
    "name": "Test Account",
    "paymentGateway": "Stripe PaymentIntents GNM Membership",
    "paymentMethod": {
      "cardNumber": "42424242424242",
      "cardType": "Visa",
      "expirationMonth": 12,
      "expirationYear": 2099,
      "secondTokenId": "test_second_token_id",
      "tokenId": "test_token_id",
      "type": "CreditCardReferenceTransaction",
    },
    "soldToContact": undefined,
  },
  "orderDate": "2025-09-17",
  "processingOptions": {
    "collectPayment": true,
    "runBilling": true,
  },
  "subscriptions": [
    {
      "customFields": {
        "DeliveryAgent__c": "",
        "ReaderType__c": "Direct",
      },
      "orderActions": [
        {
          "createSubscription": {
            "subscribeToRatePlans": [
              {
                "chargeOverrides": [
                  {
                    "pricing": {
                      "recurringFlatFee": {
                        "listPrice": 0,
                      },
                    },
                    "productRatePlanChargeId": "8ad09ea0858682bb0185880ac57f4c4c",
                  },
                ],
                "productRatePlanId": "8ad08cbd8586721c01858804e3275376",
              },
            ],
            "terms": {
              "autoRenew": true,
              "initialTerm": {
                "period": 12,
                "periodType": "Month",
                "termType": "TERMED",
              },
              "renewalSetting": "RENEW_WITH_SPECIFIC_TERM",
              "renewalTerms": [
                {
                  "period": 12,
                  "periodType": "Month",
                },
              ],
            },
          },
          "triggerDates": [
            {
              "name": "ContractEffective",
              "triggerDate": "2025-09-01",
            },
            {
              "name": "CustomerAcceptance",
              "triggerDate": "2025-09-05",
            },
          ],
          "type": "CreateSubscription",
        },
      ],
    },
  ],
}
`;
